Add geiser-kawa

spellcard199 opened the pull request:
### Brief summary of what the package does

`geiser-kawa` extends the `geiser` package with support for the Kawa Scheme language [1] [2]. 

[1] https://www.gnu.org/software/kawa/index.html
[2] https://gitlab.com/kashell/Kawa

### Direct link to the package repository

https://gitlab.com/spellcard199/geiser-kawa

### Your association with the package

Maintainer.

### Relevant communications with the upstream package maintainer

I'm the maintainer of `geiser-kawa`, but since `geiser-kawa` is an extension of the `geiser` package there has been a discussion with its maintainers about making `geiser-kawa` a separate package [1]. In summary, the `geiser` team is considering moving to an organization where there is a core `geiser` package and a `geiser-...` package for each scheme implementation so that they can focus and make more frequent updates/releases.

[1] https://gitlab.com/jaor/geiser/-/merge_requests/260

### Checklist

Please confirm with `x`:

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

### About byte-compile warnings

Result of `cask; cask build`:

```
In toplevel form:
geiser-kawa-arglist.el:110:1:Warning: Unused lexical argument ‘remote’
Compiling /home/doreen/IdeaProjects/geiser-kawa/elisp/geiser-kawa-deps.el...

In toplevel form:
geiser-kawa-deps.el:82:1:Warning: Unused lexical argument ‘desc’
geiser-kawa-deps.el:82:1:Warning: Unused lexical argument ‘buf’
Compiling /home/doreen/IdeaProjects/geiser-kawa/elisp/geiser-kawa-devutil-complete.el...
Compiling /home/doreen/IdeaProjects/geiser-kawa/elisp/geiser-kawa-devutil-exprtree.el...
Compiling /home/doreen/IdeaProjects/geiser-kawa/elisp/geiser-kawa-ext-help.el...

In toplevel form:
geiser-kawa-ext-help.el:126:1:Warning: Unused lexical argument ‘mod’
Compiling /home/doreen/IdeaProjects/geiser-kawa/elisp/geiser-kawa-globals.el...
Compiling /home/doreen/IdeaProjects/geiser-kawa/elisp/geiser-kawa-util.el...
Compiling /home/doreen/IdeaProjects/geiser-kawa/elisp/geiser-kawa.el...

In toplevel form:
geiser-kawa.el:104:1:Warning: Unused lexical argument ‘module’
```

The reason why I can't solve these "Unused lexical argument" warnings is that the culprit functions are registered to be called by `geiser` depending on the scheme implementation. In particular, this `geiser-kawa` implementation does not use those parameters, but geiser would still call those functions with that number of arguments.

### About TODOs

The reason why there are still TODOs in the code is because I don't know Kawa enough to implement them.

### About java dependencies

`geiser-kawa` is the elisp side, but the Kawa Scheme side depends on:
1. `kawa-geiser`
    - url: same repository of `geiser-kawa`
    - Takes care of wrapping required data in geiser's protocol
2. `kawa-devutil` (I'm the maintainer)
    - url: https://gitlab.com/spellcard199/kawa-devutil
    - Takes care of getting the actual data
3. `Kawa` (the actual language)
    - url: https://gitlab.com/kashell/Kawa
    - Provides the api that `kawa-devutil` uses
4. `ClassGraph`
    - url: https://github.com/classgraph/classgraph
    - Used by `kawa-devutil` for getting completion data about Java Packages and Classes

Each of these uses a GPL-Compatible Free Software License.

riscy commented:
Thanks for this, and the note about the TODOs.

> The reason why I can't solve these "Unused lexical argument" warnings

Have you tried prepending with underscores?  Just turn `variable` into `_variable`.  In the meantime, here's a very quick first pass over the files:

### geiser-kawa-arglist.el ###

byte-compile-file (using Emacs 26.3):
```
geiser-kawa-arglist.el:110:1:Warning: Unused lexical argument `remote'
```

checkdoc (using version 0.6.1):
- No issues!

package-lint-current-buffer (using version 20200419.406):
- No issues!

Suggestions/experimental static checks:
- geiser-kawa-arglist.el#L8: Prefer `https` over `http` (if possible)

### geiser-kawa-deps.el ###

byte-compile-file (using Emacs 26.3):
```
geiser-kawa-deps.el:83:1:Warning: Unused lexical argument `desc'
geiser-kawa-deps.el:83:1:Warning: Unused lexical argument `buf'
```

checkdoc (using version 0.6.1):
- No issues!

package-lint-current-buffer (using version 20200419.406):
- No issues!

Suggestions/experimental static checks:
- geiser-kawa-deps.el#L8: Prefer `https` over `http` (if possible)

### geiser-kawa-devutil-complete.el ###

byte-compile-file (using Emacs 26.3):
- No issues!

checkdoc (using version 0.6.1):
- No issues!

package-lint-current-buffer (using version 20200419.406):
- No issues!

Suggestions/experimental static checks:
- geiser-kawa-devutil-complete.el#L8: Prefer `https` over `http` (if possible)
- geiser-kawa-devutil-complete.el#L142: You can use `not` or `null`

### geiser-kawa-devutil-exprtree.el ###

byte-compile-file (using Emacs 26.3):
- No issues!

checkdoc (using version 0.6.1):
- No issues!

package-lint-current-buffer (using version 20200419.406):
- No issues!

Suggestions/experimental static checks:
- geiser-kawa-devutil-exprtree.el#L8: Prefer `https` over `http` (if possible)

### geiser-kawa-ext-help.el ###

byte-compile-file (using Emacs 26.3):
```
geiser-kawa-ext-help.el:126:1:Warning: Unused lexical argument `mod'
```

checkdoc (using version 0.6.1):
- No issues!

package-lint-current-buffer (using version 20200419.406):
- No issues!

Suggestions/experimental static checks:
- geiser-kawa-ext-help.el#L99: No `format` required; messages are already f-strings
- geiser-kawa-ext-help.el#L121: No `format` required; messages are already f-strings
- geiser-kawa-ext-help.el#L8: Prefer `https` over `http` (if possible)

### geiser-kawa-globals.el ###

byte-compile-file (using Emacs 26.3):
- No issues!

checkdoc (using version 0.6.1):
- No issues!

package-lint-current-buffer (using version 20200419.406):
- No issues!

Suggestions/experimental static checks:
- geiser-kawa-globals.el#L8: Prefer `https` over `http` (if possible)

### geiser-kawa-util.el ###

byte-compile-file (using Emacs 26.3):
- No issues!

checkdoc (using version 0.6.1):
- No issues!

package-lint-current-buffer (using version 20200419.406):
```
1 issue found:
68:0: error: `geiser-kawa-util--eval/result' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
```

Suggestions/experimental static checks:
- geiser-kawa-util.el#L39: Consider `unless ...` instead of `when (not ...)`
- geiser-kawa-util.el#L8: Prefer `https` over `http` (if possible)

### geiser-kawa.el ###

byte-compile-file (using Emacs 26.3):
```
geiser-kawa.el:104:1:Warning: Unused lexical argument `module'
```

checkdoc (using version 0.6.1):
- No issues!

package-lint-current-buffer (using version 20200419.406):
- No issues!

Suggestions/experimental static checks:
- geiser-kawa.el#L66: It's safer to sharp-quote function names; use `#'`
- geiser-kawa.el#L9: Prefer `https` over `http` (if possible)

<!--
### Loadability ###

Running #'load on each file:
```
Loading geiser-kawa-arglist.el
Loading geiser-kawa-deps.el
Loading geiser-kawa-devutil-complete.el
Loading geiser-kawa-devutil-exprtree.el
Loading geiser-kawa-ext-help.el
Loading geiser-kawa-globals.el
Loading geiser-kawa-util.el
Loading geiser-kawa.el
Done.
```
-->

### Packaging ###

<!-- ```(geiser-kawa
 :fetcher gitlab
 :repo "spellcard199/geiser-kawa"
 :files ("elisp/*.el"
         "pom.xml"
         ".mvn" "mvnw" "mvnw.cmd"
         "src"))``` -->

<!-- COPYING excerpt: `BSD 3-Clause License...` -->

A note about your license boilerplate in each file -- you have written "This program is free software" but the term "free software" is (usually) in reference to GPL-licensed software.  Recently I've been recommending using an [SPDX-License-Identifier](https://spdx.org/using-spdx-license-identifier) maybe e.g. under your `Package-Requires` line.  Those license identifiers are machine detectable, unambiguous, and terse.

spellcard199 commented:
Thank you for your suggestions. The following issues should be fixed now.

```
geiser-kawa-arglist.el:110:1:Warning: Unused lexical argument ‘remote’
geiser-kawa-deps.el:82:1:Warning: Unused lexical argument ‘desc’
geiser-kawa-deps.el:82:1:Warning: Unused lexical argument ‘buf’
geiser-kawa-ext-help.el:126:1:Warning: Unused lexical argument ‘mod’
geiser-kawa.el:104:1:Warning: Unused lexical argument ‘module’
geiser-kawa-devutil-complete.el#L142: You can use not or null
geiser-kawa-ext-help.el#L99: No format required; messages are already f-strings
geiser-kawa-ext-help.el#L121: No format required; messages are already f-strings
68:0: error: `geiser-kawa-util--eval/result' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
geiser-kawa-util.el#L39: Consider unless ... instead of when (not ...)
geiser-kawa.el#L66: It's safer to sharp-quote function names; use #'
```

> I've been recommending using an SPDX-License-Identifier maybe e.g. under your Package-Requires line.

The license notice has been replaced with `SPDX-License-Identifier: BSD-3-Clause` in each file. Is this ok or should I put it just in `geiser-kawa.el` and remove it from the other files?

I also have 2 questions relative to the `Suggestions/experimental static checks`:
1. Are there instructions on how to run these static checks myself?
2. I have no idea about what all those `Prefer https over http (if possible)` are referring to. What do I do?

spellcard199 commented:
> I have no idea about what all those `Prefer https over http (if possible)` are referring to. What do I do?
Answering myself: probably they were referring to the URL in the license notice I replaced with the SPDX-License-Identifier.

Now `cask clean-elc; cask build` produces no warnings so I updated the checklist. However, I still haven't found out how to check for `Suggestions/experimental static checks` myself.

riscy commented:
Apologies for the backlog.  

> Is this ok or should I put it just in geiser-kawa.el and remove it from the other files?

It should be in all the files (but it's just one line, which is the nice thing).

> I still haven't found out how to check for Suggestions/experimental static checks myself.

These come from [a handful of scripts I run](http://github.com/riscy/melpazoid) to help automate checking these PR's.  There are a couple ways to use it (see the README) and it can be incorporated into CI, but I haven't learned how GitLab's builtin CI works yet.  But there are [instructions for other tools](https://github.com/riscy/melpazoid#usage).

riscy commented:
The checklist looks great.  Taking a deeper look --

- I'm not familiar with `geiser-impl--add-to-alist` but usually when you add to auto-mode-alists the regular expression should resemble, e.g., `\\.scm\\'` rather than `\\.scm$` by convention.
- I think the coupling of the java and maven build scripts with the elisp is fine (other packages do this), but I thought I'd double check with you that these various source files are intrinsically linked (and there's not, e.g., a situation where a user would want to update the java files but not the el files, etc.).  Apologies for my ignorance on that one.
- In geiser-kawa-arglist.el I see you call process-lines and I'd recommend taking a peek at the `shell-quote-argument` function if you haven't considered it (e.g., could geiser-kawa-deps-jar-path have spaces in it, no matter how far-fetched the idea is?  worth handling?)
- `geiser-kawa-deps--run-kawa--compil-hook` -> compile-hook

Beyond these I think this is ready to merge.

spellcard199 commented:
Thank you for your time and work.

> * I'm not familiar with `geiser-impl--add-to-alist` but usually when you add to auto-mode-alists the regular expression should resemble, e.g., `\\.scm\\'` rather than `\\.scm$` by convention.

Tested and works as well. Fixed.

> * I think the coupling of the java and maven build scripts with the elisp is fine (other packages do this), but I thought I'd double check with you that these various source files are intrinsically linked (and there's not, e.g., a situation where a user would want to update the java files but not the el files, etc.).  Apologies for my ignorance on that one.

In _my_ ignorance I didn't even think about it. I thought the user could just re-install the whole thing, since:
- Names, parameters and return value of the scheme procedures that elisp calls must be "in sync" between elisp and scheme
- Tests can't catch the case where the user updated just the java part and some procedure changed (e.g.) the "shape" of the returned list

If it's preferable, it is possible to split the project into separate repositories, depending on what you are willing to give up.

Beside the elisp directory the melpa recipe also includes:
- `mvnw`, `mvnw.cmd`, `.mvn`: I didn't write these. They are added by [maven wrapper](https://github.com/takari/maven-wrapper) using `mvn -N io.takari:maven:.....:wrapper`
- `src` and `pom.xml`: java counterpart of `geiser-kawa` (uses Maven as build tool)

Q. What does a user lose removing `mvnw`, `mvnw.cmd`, `.mvn`?
A. Ability to auto-download Maven: user would need to download Maven her/himself.

Q. What does a user lose moving to another repository both `src` and `pom.xml` and replacing them with a `pom.xml` that depends on them?
A. Just the ability to read and customize the java code without setting up a new project. No additional work becomes required.

Q. What does a user lose moving to another repository both `src` and `pom.xml` without replacing them with another `pom.xml` that depends on them?
A. Ability to use just Maven to download the (moved) java dependency, since maven needs a `pom.xml` as starting point. Probably user would also need to have `git` installed.

> * In geiser-kawa-arglist.el I see you call process-lines and I'd recommend taking a peek at the `shell-quote-argument` function if you haven't considered it (e.g., could geiser-kawa-deps-jar-path have spaces in it, no matter how far-fetched the idea is?  worth handling?)

I just tried adding `(shell-quote-command ... )` in various places and it invariably breaks the function. Things I have tried:
- replacing the body of `geiser-kawa-arglist--make-classpath-arg`:
    - from:
    `(format "-Djava.class.path=%s" classpath)`
    - to:
    `(format "-Djava.class.path=%s" (shell-quote-command classpath))`
    - or even to:
    `(format "-Djava.class.path=\"%s\"" classpath)`
    - or even to:
    `(shell-quote-command (format "-Djava.class.path=%s" classpath))`
- replacing the line containing `process-lines`:
    - from:
    ` (output (apply #'process-lines (cons program args)))`
    - to:
    ```
    (output (apply #'process-lines
                    (cons program
                          (mapcar
                            (lambda (arg) (shell-quote-argument arg))
                            args))))
    ```

While I don't understand why, as it is now it works even if there are spaces in `geiser-kawa-deps-jar-path` (tried on linux).

> * `geiser-kawa-deps--run-kawa--compil-hook` -> compile-hook

I just added the `e`.

riscy commented:
> Names, parameters and return value of the scheme procedures that elisp calls must be "in sync" between elisp and scheme ...

That makes sense and is a great reason to package them together -- thanks for your well-reasoned response.  Regarding the quoting, I figure since the values are fully under the user's control it's fine.

One last suggestion --

- geiser-kawa.el#L62: It's safer to sharp-quote function names; use `#'`

but extremely minor (and not a blocker) and from here this looks good to merge.  And, welcome to MELPA! :)

spellcard199 commented:
> * geiser-kawa.el#L62: It's safer to sharp-quote function names; use `#'`

Next thing I'm going to learn will be how to use your scripts. Fixed.

> And, welcome to MELPA! :)

Thanks!
